### PR TITLE
[RN] Add Android and iOS SDKs

### PR DESCRIPTION
**Pull request by saghul**
More details on each individual commit.

The API is now ready for review on both platforms.

#### Still TODO

- [x] android: docs
- [x] android: bundle JS automagically
- [x] android: TEST TEST TEST

### CODE REVIEW

**Comment by saghul** [comment]
@lyubomir I added the alternative implementation on iOS. It looks much nicer, we should totally go that route.

**Comment by saghul** [comment]
New Android API using a view with FrameLayout. If we agree on this being the way to go, I'll squash these 2 commits with the others, not to do it in 2 steps.

File: android/app/src/main/java/org/jitsi/meet/MainActivity.java [hunk 1]
```diff
@@ -1,46 +1,98 @@
...
-    protected String getMainComponentName() {
-        return "App";
+    public void onNewIntent(Intent intent) {
+        JKConferenceView.onNewIntent(intent);
     }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        JKConferenceView.onHostDestroy(this);
```
Comment (lyubomir) [inline android/app/src/main/java/org/jitsi/meet/MainActivity.java]: I do not understand why there are multiple static methods on `JKConferenceView` and I don't think that the javadocs on them really explain to me as a consumer of the API why I need to call them. My ideal Android View will not require such methods to be called by its consumer. If it's all about deep linking, cannot we extract linking in a separate API like React Native's LInking? I hardly think a non-trivial app will embed a Jitsi Meet view and agree for the app's linking to be entirely handled by the Jitsi Meet view.

File: android/app/src/main/java/org/jitsi/meet/MainActivity.java [hunk 2]
```diff
@@ -1,46 +1,98 @@
...
+ * "singleTask" mode, so it will be created upon application initialization and there will be
+ * a single instance of it. Further attempts at launching the application once it was already
+ * launched will result in <tt>onNewIntent</tt> being called.
+ *
+ * In this activity we use a single <tt>JitsiConference</tt> instance. This instance gives us
+ * access to a view which displays the welcome page and the conference itself. All lifetime methods
+ * associated with this Activity are hooked to the React Native subsystem via proxy calls through
+ * the <tt>JitsiMeet</tt> singleton.
+ */
+public class MainActivity extends AppCompatActivity {
```
Comment (lyubomir) [inline android/app/src/main/java/org/jitsi/meet/MainActivity.java]: Once the Android View is ready, I think it'd be nice to add an Activity to the Android SDK as well because that's another popular Android "way".

Comment (saghul) [inline android/app/src/main/java/org/jitsi/meet/MainActivity.java]: Agreed!

File: android/app/src/main/java/org/jitsi/meet/MainActivity.java (hunk 1 above)
Comment (saghul) [inline android/app/src/main/java/org/jitsi/meet/MainActivity.java]: The docs got scrapped when I added the view, I'll brind those back when I merge the approaches. In a nutshell, this is not about deep linking, but about activity management. Those activity lifecycle methods need to be tied to React Native.

I'm happy to move those elsewhere, but they need to call stuff on the React native Instance Manager, of which we hold a single instance. That's why I put them in the view as static methods.

I'll make this clear in the docs.

File: android/app/src/main/java/org/jitsi/meet/MainActivity.java [hunk 3]
```diff
@@ -1,46 +1,98 @@
...
+import android.os.Build;
+import android.provider.Settings;
+import android.support.v7.app.AppCompatActivity;
+import android.os.Bundle;
+
+import org.jitsi.jitsikit.*;
+
+
+/**
+ * The one and only Activity that Jitsi Meet (the app) needs. The activity is launched in
```
Comment (saghul) [inline android/app/src/main/java/org/jitsi/meet/MainActivity.java]: Sure. I didn't pay much attention to that when doing the view part, I'll fix those uop ASAP.

**Comment by saghul** [comment]
About naming: we seem to have started to use "Jitsi" as the umbrella name for everything: https://jitsi.org/what-is-jitsi/ hence I chose "JitsiKit" on iOS, and the JKSomethingSomething convention, following other frameworks in the Apple echosystem. 

I then translated those to Android. It looks a bit alien at first, so I'm happy to adjust there.

At any rate, changing names is easy to do. I'll merge both approaches into single commits and fix the docs, then we can settle on the name.

File: ios/sdk/JKConferenceViewDelegate.h [hunk 4]
```diff
@@ -0,0 +1,6 @@
+
+@protocol JKConferenceViewDelegate <NSObject>
+
+// TODO: add delegate methods
```
Comment (saghul) [inline ios/sdk/JKConferenceViewDelegate.h]: Correct. The events will be added as a separate commit though, if that's ok.

**Comment by lyubomir** [comment]
@saghul, when I build the iOS app, I'm seeing signing steps for WebRTC.framework and JitsiMeet.framework. If that only happens for the app, that'd be fine. But I don't know of a way to build JitsiMeet.framework without building the app so my concern about signing JitsiMeet.framework arrises. Firstly, signing WebRTC.framework is a little bit futile because we do not sign it in react-native-webrtc from where jitsi-meet gets it. Secondly and most importantly, signing JitsiMeet.framework is an unanswered question at this point because there's no ready way to get the built JitsiMeet.framework from where we build the app for distribution in the App Store or for enterprise distribution. So I only know to build it on my computer but then the signing material will be useless to other developers. Could you please shed some light on the subject?

**Comment by saghul** [comment]
@lyubomir 

> when I build the iOS app, I'm seeing signing steps for WebRTC.framework and JitsiMeet.framework.

Yep, that's right. Let me explain: frameworks are supposed to be built unsigned, and then when an app embeds them it will sign them with the same signature as the app itself. Now, here I ran into a little problem: Apple discourages (but doesn't forbid) frameworks from embedding other frameworks. Part of this discouragement results in nested frameworks not being signed. That is, if we don't sign JitsiMeet.framework, WebRTC.framework won't be signed automagically when Jitsi Meet (the app) signs JitsiMeet.framework. There are a number of ways around this, I chose to force signing for us, since I ran into some problems because I have more than one Apple Developer account. This should not be a problem in practice, since the machines which do the signing will not have that.

> But I don't know of a way to build JitsiMeet.framework without building the app so my concern about signing JitsiMeet.framework arrises.

I'll add a scheme which builds JitsiMeet.framework without any signing. This is what we will distribute, or alternatively we can sign it too, with the release key in the release machine, though that would be a bit unpractical for development.

> So I only know to build it on my computer but then the signing material will be useless to other developers. Could you please shed some light on the subject?

The framework should go unsigned, and it will be signed automatically when the app embeds it. There is a manual step required so the embedded WebRTC framework is also signed, since Xcode won't do this automatically.

Apple's suggestion is that the app should link directly with all frameworks, so no framework would embed another. I don't like this and since there is a way to make it work, I went there. I see no reason why Jitsi Meet (the app) should link with WebRTC.framework, it should only know about JitsiMeet.framework.

Now, all that said, one thing I want to look into for the future is making a static build instead of a framework build in react-native-webrtc, that would simplify this process, and might be a better idea overall.

**Comment by saghul** [comment]
@lyubomir Here is how you build the framework alone:

First change the target and select "JitsiMeet".

screen shot 2017-06-07 at 11 02 45

Make sure you select "Generic iOS device" and hit run.

screen shot 2017-06-07 at 11 03 00

Then go to the files view, and inside "Products" right click and pick "Show in Finder".

screen shot 2017-06-07 at 11 03 13

**Comment by saghul** [comment]
@lyubomir I took another look at signing the embedded frameworks and this time I prevailed! 

I modified JitsiMeet.framework not to be signed (as Apple likes it) and then added an extra codesign step in the fixup-frameworks script (see the last commit).

Since any application using JitsiMeet.framework will need a similar script (to strip simulator architectures for Store submission) I think this is a good and pretty frictionless solution.

File: ios/sdk/src/JitsiMeetView.m [hunk 5]
```diff
@@ -136,6 +141,20 @@ - (void)loadURL:(NSURL *)url {
...
+     * bridge. Once we have that sorted out multiple instances of JitsiMeetView
+     * will be allowed.
+     */
+    if (instance != nil) {
+        @throw [NSException
+            exceptionWithName:@"RuntimeError"
+                       reason:@"Only a single instance is currently allowed"
+                     userInfo:nil];
+    }
+    instance = self;
```
Comment (lyubomir) [inline ios/sdk/src/JitsiMeetView.m]: On both Android and iOS you only ever assign non-`null` to the static `mInstance` and `instance`, respectively. Consequently, that precludes the scenario of 2 `JitsiMeetView`s which do NOT exist simultaneously.

While I do understand that it's very unlikely that the scenario with 2 existing `JitsiMeetView`s which simultaneously drive 2 conferences will work because of non-shareable resources which we cannot make shareable, I think the scenario of multiple non-simultaneous `JitsiMeetView`s is likely.

File: react/features/mobile/external-api/middleware.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ * @param {Store} store - Redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(store => next => action => {
+    const eventData = {};
+
+    switch (action.type) {
+    case CONFERENCE_FAILED: {
+        eventData.error = action.error;
+        eventData.url = getInviteURL(store.getState());
```
Comment (lyubomir) [inline react/features/mobile/external-api/middleware.js]: 1. `inviteURL` in pretty much meant for display and/or public purposes and, consequently, it doesn't contain URL `hash` and `search` params. Which to me makes it kind of less useful to Jitsi Meet SDK consumers because they specify a URL to `JitsiMeetView.loadURL(URL)` but then potentially get another URL in `JitsiMeetView`'s delegate/listener. `locationURL` is closer to what I'd go with.

2. Technically, there's a possibility that Jitsi Meet will rewrite the URL specified to `loadURL`. It may replace the scheme (if the specified URL uses the app-specific scheme), infer other URL properties from the app's `defaultURL`, and plainly replace the `host` and `pathname` for certain hardcoded `host`s.

3. Due to specifics of the current implementation of tapping on the hangup `ToolbarButton` (which are you painfully aware of), `CONFERENCE_WILL_LEAVE` and `CONFERENCE_LEFT` will report an `eventData.url` which is the app's `defaultURL`.

(4. I hope the type of the value of `action.error` is really a `string`. I know the `action`'s `type` is documented like that but it's not really enforced. If it's not a `string`, I'm not sure how useful that would be to the Jitsi Meet SDK consumers.)

Hopefully, the fact that `eventData` is a dictionary/hashmap for Jitsi Meet SDK consumers discourages them from examining it at all.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    public static void onNewIntent(Intent intent) {
+        if (mReactInstanceManager != null) {
+            mReactInstanceManager.onNewIntent(intent);
+        }
+    }
+
+    /**
+     * Interface for listening to events coming from Jitsi Meet.
+     */
+    public interface Listener {
```
Comment (lyubomir) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java]: It'd be nice to Jitsi Meet SDK for Android consumers if we had a default/adapter implementation of `JitsiMeetView.Listener`. Otherwise, they'd have to (1) implement all methods even if they were interested in one of them only (which is our hope at this time of writing if I understand you correctly) and/or (2) amend their source code every time we added a new method to the interface even if they weren't interested in it.

(That's one more advantage to having a 1-method interface in the face of the `HashMap`s.)

**Comment by lyubomir** [comment]
Merged into jitsi/jitsi-meet#master as https://github.com/jitsi/jitsi-meet/commit/e5cc8cd32b8f885a9ce07bb639fed03823d97b6a.

**Comment by saghul** [comment]
> (4. I hope the type of the value of action.error is really a string. I know the action's type is documented like that but it's not really enforced. If it's not a string, I'm not sure how useful that would be to the Jitsi Meet SDK consumers.)

Yes, it's a string we pass from JS. Its usefulness if debatable for purposes other than logging, I agree.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java (hunk 7 above)
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java]: I can add a JitsiMeetViewAbstractListener class with all methods raising UnsupportedOperationException. WDYT?

Comment (lyubomir) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java]: If all methods raise `UnsupportedOperationException`, then the implementers are better off without it (because they'll at least suffer the performance penalty incurred by throwing exceptions). If all methods are empty/do nothing, then it's a help because the implementers can override whichever they are interested in. Additionally, "empty methods" is what you gave iOS developers.

**Comment by saghul** [comment]
> If all methods raise UnsupportedOperationException, then the implementers are better off without it (because they'll at least suffer the performance penalty incurred by throwing exceptions). If all methods are empty/do nothing, then it's a help because the implementers can override whichever they are interested in.

Ok, I didn't know about that. Should I remove the handling of UnsupportedOperationException then?

> Additionally, "empty methods" is what you gave iOS developers.

Not quite. Delegate methods can be marked optional, but there is no equivalent in a Java interface AFAIK. So the abstract class sounds like the way to go 👍
